### PR TITLE
Add Terraformer instance type and other modifications for RTA

### PR DESCRIPTION
**Pull request by jsf9k**
## 🗣 Description ##

This pull request adds Terraformer instances to this repository and makes assorted other changes for RTA.

## 💭 Motivation and context ##

See cisagov/cool-system-internal#5.

## 🧪 Testing ##

I used these changes to successfully spin up a Teamserver (using the Terraformer role created herein).  All changes are already applied to env3 in our staging COOL environment.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: terraformer_cloud_init.tf
Comment (dav3r) [inline terraformer_cloud_init.tf]: The VNC user's username should be pulled in from the SSM parameter store (see `var.ssm_key_vnc_username`).

File: terraformer_ec2.tf
Comment (dav3r) [inline terraformer_ec2.tf]: It sounds like this should be a `TODO` with a new issue to go along with it.

File: terraformer_policy.tf
Comment (dav3r) [inline terraformer_policy.tf]: Shouldn't this comment be written more generically, since we are comparing against the value of `var.tags["Team"]` (which may or may not be "VM Fusion - Development")?

File: terraformer_role.tf
Comment (dav3r) [inline terraformer_role.tf]: As above, this comment should be more generic.

**Review by dav3r — APPROVED** [review]
This looks excellent to me, though I did note a few small items for you to take a look at. 🏗️

File: terraformer_role.tf
Comment (jsf9k) [reply terraformer_role.tf]: See commit acde207.

File: terraformer_policy.tf
Comment (jsf9k) [reply terraformer_policy.tf]: See commit acde207.

Comment (dav3r) [reply terraformer_policy.tf]: That works for me, thanks! 👍

File: terraformer_cloud_init.tf
Comment (jsf9k) [reply terraformer_cloud_init.tf]: See commit f146058.

Comment (dav3r) [reply terraformer_cloud_init.tf]: That's the stuff right there! 🪐

File: terraformer_ec2.tf
Comment (jsf9k) [reply terraformer_ec2.tf]: See commit 5bc0ddb.

File: outputs.tf
Comment (dav3r) [inline outputs.tf]: I think we have standardized on the capitalization below in this this repo:

```suggestion
  description = "The instance profile for the Debian desktop instances."
```

File: terraformer_iam.tf
Comment (dav3r) [inline terraformer_iam.tf]: Missing a word:
```suggestion
# Allow the Terraformer instance to assume the necessary role to
```

**Review by dav3r — APPROVED** [review]
Outstanding work on this PR - everything looks great! 🏋️

**Review by dav3r — APPROVED** [review]
Still looks great! 🦚 

The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

**Comment by jsf9k** [comment]
> Still looks great! 
> 
> The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

When you say "aliasing" do you mean "symlinking"?

**Comment by dav3r** [comment]
> > Still looks great!
> > The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).
> 
> When you say "aliasing" do you mean "symlinking"?

Yes, that is indeed what I meant.

**Comment by jsf9k** [comment]
> Still looks great! 
> 
> The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

Check out commit c4e1c1a.

**Comment by dav3r** [comment]
> > Still looks great!
> > The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).
> 
> Check out commit [c4e1c1a](https://github.com/cisagov/cool-assessment-terraform/commit/c4e1c1ad273c2064da175c2f263d92638a7b95b7).

Lovely! 🚀

**Review by dav3r — APPROVED** [review]

File: cloud-init/write-aws-config.tpl.sh
Comment (dav3r) [inline cloud-init/write-aws-config.tpl.sh]: Is it worth generalizing this variable so that it is not so explicitly tied to RTA and their redirector account?  The answer may be "no", but I wanted to make sure we at least discussed it.  I don't have a great name suggestion, but I was thinking something along the lines of `assessor_owned_route53_role_arn` - that can be a starting point - I'm sure we can do better.

Comment (jsf9k) [reply cloud-init/write-aws-config.tpl.sh]: Really I'm just creating the profile.  If you have a role from any other account for which you have delegated authority to this assessment account then you could put it there.  The RTA folks just happen to need to give us a role to allow them to make Route53 changes in their own account.  In that sense the profile is very general and I could just as well call it `role_arn`.  We won't know or care what they use the profile to do.

That said, I hadn't really envisioned any other uses for this role.  I guess I'm fine with changing the name if you and @mcdonnnj can decide on what you want.

Comment (dav3r) [reply cloud-init/write-aws-config.tpl.sh]: The general nature of this role makes it tough to name, but I suppose as long as we document it well enough then that won't matter.  I'd be fine with explicitly referencing this RTA/Route53 use case in the variable description.  What do you think about `assessment_team_delegated_account_role_arn`?

Comment (jsf9k) [reply cloud-init/write-aws-config.tpl.sh]: I ended up going with `assessor_account_role_arn` as I felt the rest was redundant.  See commit 65b1e21.

File: README.md
Comment (dav3r) [inline README.md]: You mixed up the variable name here:
```suggestion
| assessor\_account\_role\_arn | The ARN of an IAM role that can be assumed to create, delete, and modify AWS resources in a separate assessor-owned AWS account. | `string` | `"arn:aws:iam::123456789012:role/Allow_It"` | no |
```

**Review by dav3r — APPROVED** [review]
Aside from the small README item I just mentioned, this all looks great! 🐲

File: cloud-init/write-aws-config.tpl.sh
Comment (dav3r) [reply cloud-init/write-aws-config.tpl.sh]: `assessor_account_role_arn` works for me - thanks!

File: terraformer_sg.tf
Comment (mcdonnnj) [inline terraformer_sg.tf]:
```suggestion
# For: Terraformer instances must be able to terraform init.
```

**Review by dav3r — APPROVED** [review]
👍 👍

File: .github/workflows/build.yml
Comment (dav3r) [inline .github/workflows/build.yml]: This change is reverted in https://github.com/cisagov/cool-assessment-terraform/pull/136, but just wanted to mention it to make sure that it flows into this PR.

File: terraformer_sg.tf
Comment (dav3r) [inline terraformer_sg.tf]: Is there a reason this block is not at the bottom of this file?  It breaks the order of our rules being sorted by port number.

**Review by dav3r — APPROVED** [review]
Approved, with a couple of comments. 👍

File: terraformer_sg.tf
Comment (jsf9k) [reply terraformer_sg.tf]: I was thinking that the `ssh` and `winrm` jazz should live together since it serves the same purpose.  You're argument is just as valid, though, so I'll move it.

Comment (jsf9k) [reply terraformer_sg.tf]: See commit 200ce1b.

Comment (dav3r) [reply terraformer_sg.tf]: I knew there was some method to your madness, but I wasn't sure what it was.  Thanks for moving that block. 👍

File: .github/workflows/build.yml
Comment (jsf9k) [reply .github/workflows/build.yml]: Yep.  Taken care of in commit 1110e86.

File: .pre-commit-config.yaml
Comment (mcdonnnj) [inline .pre-commit-config.yaml]: I no longer see this directory. Is this still needed?

```suggestion
```

File: cloud-init/write-aws-config.tpl.sh
Comment (mcdonnnj) [inline cloud-init/write-aws-config.tpl.sh]: Are we unable to use the long flags here?

```suggestion
mkdir --parents "$d"
chown --recursive "$vnc_username:$vnc_username" "$d"
```

**Comment by jsf9k** [comment]
The checks should pass again once the dust settles on our Terraform upgrade.  I will re-run the checks once we're in a good state there.  I'm also waiting until then to attempt to apply the changes from the most recent commit to an environment in our staging COOL environment.

File: outputs.tf
Comment (mcdonnnj) [inline outputs.tf]: Is there a reason this output is named differently from the previous two for the same group of resources?

```suggestion
output "assessor_portal_security_group" {
```

File: outputs.tf
Comment (mcdonnnj) [inline outputs.tf]: Should this be a proper noun?

```suggestion
  description = "The AWS EC2 instance hosting Guacamole."
```

File: outputs.tf
Comment (mcdonnnj) [inline outputs.tf]: Should these all be prefixed `pentest_portal_*`? That would be in line with 2/3 `assessor_portal_*` outputs and all `debian_desktop_*` outputs previously declared.

File: private_acl_rules.tf
Comment (mcdonnnj) [inline private_acl_rules.tf]:
```suggestion
# (i.e. Guacamole, Mattermost, etc.)
```

File: private_acl_rules.tf
Comment (mcdonnnj) [inline private_acl_rules.tf]: It might just be me but there's enough going on here that a brief comment might be good for future skimmers.

File: providers.tf
Comment (mcdonnnj) [inline providers.tf]:
```suggestion
# for this Terraform root module's Terraform state.
```

File: read_terraform_state_role.tf
Comment (mcdonnnj) [inline read_terraform_state_role.tf]:
```suggestion
    # do have to put something there as a placeholder, though.
```

File: terraformer_assume_role_policy_doc.tf
Comment (mcdonnnj) [inline terraformer_assume_role_policy_doc.tf:9]:
```suggestion
      "sts:AssumeRole",
      "sts:TagSession",
```

File: outputs.tf
Comment (mcdonnnj) [inline outputs.tf:201]: Just wanted to verify that there's no `terraformer_instance_profile` output by design.

File: terraformer_iam.tf
Comment (mcdonnnj) [inline terraformer_iam.tf]: Should this name include "terraformer" to be specific?

File: operations_acl_rules.tf
Comment (mcdonnnj) [inline operations_acl_rules.tf]:
```suggestion
# public access to Cobalt Strike Teamservers on port 50050.
```

File: operations_acl_rules.tf
Comment (mcdonnnj) [inline operations_acl_rules.tf]:
```suggestion
# public access to WinRM on port 5986 or Cobalt Strike Teamservers on
```

File: operations_acl_rules.tf
Comment (mcdonnnj) [inline operations_acl_rules.tf]:
```suggestion
# For: Assessment team VNC access from private subnet to operations.
```

File: examples/use-terraformer-instance/README.md
Comment (mcdonnnj) [inline examples/use-terraformer-instance/README.md]: Are we intentionally leaving out a `versions.tf` file for this example?

File: examples/use-terraformer-instance/README.md
Comment (mcdonnnj) [inline examples/use-terraformer-instance/README.md]:
```suggestion
| tags | Tags to apply to all AWS resources created. | `map(string)` | `{}` | no |
```

File: examples/use-terraformer-instance/README.md
Comment (mcdonnnj) [inline examples/use-terraformer-instance/README.md]: Should this content just be removed since this is an example in a parent repository?

File: examples/use-terraformer-instance/outputs.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/outputs.tf]:
```suggestion
  description = "The Teamserver instance."
```

File: examples/use-terraformer-instance/outputs.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/outputs.tf]:
```suggestion
  description = "The Teamserver A record."
```

File: examples/use-terraformer-instance/outputs.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/outputs.tf]:
```suggestion
  description = "The Teamserver EIP."
```

File: examples/use-terraformer-instance/teamserver_cloud_init.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/teamserver_cloud_init.tf]:
```suggestion
  # Add https-certificate section to Cobalt Strike profiles
```

File: examples/use-terraformer-instance/teamserver_ec2.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/teamserver_ec2.tf]:
```suggestion
# The Elastic IP for the Teamserver
```

File: examples/use-terraformer-instance/teamserver_ec2.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/teamserver_ec2.tf]:
```suggestion
# The EIP association for the Teamserver
```

File: examples/use-terraformer-instance/teamserver_ec2.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/teamserver_ec2.tf]:
```suggestion
# The Teamserver EC2 instances
```

File: examples/use-terraformer-instance/teamserver_ec2.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/teamserver_ec2.tf]:
```suggestion
# The Teamserver AMI
```

File: examples/use-terraformer-instance/variables.tf
Comment (mcdonnnj) [inline examples/use-terraformer-instance/variables.tf]:
```suggestion
  description = "Tags to apply to all AWS resources created."
```

File: examples/use-terraformer-instance/README.md
Comment (mcdonnnj) [inline examples/use-terraformer-instance/README.md]:
```suggestion
| teamserver | The Teamserver instance. |
| teamserver\_a\_record | The Teamserver A record. |
| teamserver\_eip | The Teamserver EIP. |
```

**Review by mcdonnnj — APPROVED** [review]
Thank you for your work on getting this functionality built out. Some strong work 💪💪💪. I have a bevy of suggestions relating to capitalization and name consistency. There are a few other questions and suggestions smattered amongst these.

File: examples/use-terraformer-instance/README.md
Comment (jsf9k) [reply examples/use-terraformer-instance/README.md]: Please see commit a32d80f.

File: examples/use-terraformer-instance/README.md
Comment (jsf9k) [reply examples/use-terraformer-instance/README.md]: Please see commit e595ead.

File: outputs.tf
Comment (jsf9k) [reply outputs.tf]: See commit b7a7c97.

File: outputs.tf
Comment (jsf9k) [reply outputs.tf:201]: It is by design.  That's a powerful instance profile, and I thought it best if it wasn't dangling out there tempting folks to make use of it.

File: outputs.tf
Comment (jsf9k) [reply outputs.tf]: Please see commit fc8e68b.

File: private_acl_rules.tf
Comment (jsf9k) [reply private_acl_rules.tf]: Please see commit 68bf56a.

File: terraformer_iam.tf
Comment (jsf9k) [reply terraformer_iam.tf]: Good call.  Please see commit 65663d2.

File: examples/use-terraformer-instance/variables.tf
Comment (jsf9k) [reply examples/use-terraformer-instance/variables.tf]: Please see commit 42449436df2217a1e51fae68e2ff686eabc3a1d8.

File: examples/use-terraformer-instance/README.md
Comment (jsf9k) [reply examples/use-terraformer-instance/README.md]: Please see commit 42449436df2217a1e51fae68e2ff686eabc3a1d8.

File: outputs.tf
Comment (jsf9k) [reply outputs.tf]: Please see commit a22bb6456830b2526ec2ff47b45aa94ba1435e80.

**Comment by jsf9k** [comment]
@mcdonnnj can you circle back and revisit the unresolved comments when you have a chance?  I believe I have taken care of all of them.

**Comment by jsf9k** [comment]
> @mcdonnnj can you circle back and revisit the unresolved comments when you have a chance? I believe I have taken care of all of them.

Thank you good sir!

**Comment by jsf9k** [comment]
I will wait until #137 is merged before merging this PR.

File: examples/use-terraformer-instance/README.md
Comment (dav3r) [inline examples/use-terraformer-instance/README.md]: To reflect the change in https://github.com/cisagov/cool-assessment-terraform/pull/132/commits/54f4e06f322ebb450a768a9accc5480f2025309f:
```suggestion
| terraform | ~> 0.14.0 |
```

**Review by dav3r — APPROVED** [review]
Everything looks groovy, baby, yeah! 🇬🇧

File: locals.tf
Comment (mcdonnnj) [inline locals.tf:157]: So while re-reviewing I noticed that `ipa_ports` (above) uses `proto` but this map uses `protocol`. I think we should standardize against one or the other to keep things consistent (`ipa_ports` also doesn't have sorted keys but that's a thing for another time).

**Review by mcdonnnj — APPROVED** [review]
Still looking good ✨. I had one question/suggestion around consistent naming conventions.

File: locals.tf
Comment (jsf9k) [reply locals.tf:157]: See commit e0cd6a2.

**Pull request merged**
